include/af/memory.h

```diff
@@ -0,0 +1,339 @@
+    buffer_info(void* ptr, af_event event);
+    ~buffer_info();
+#if AF_COMPILER_CXX_RVALUE_REFERENCES
+    buffer_info(buffer_info&& other);
+    buffer_info& operator=(buffer_info&& other);
+#endif
+    void* getPtr() const;
+    void setPtr(void* ptr);
+    af_event getEvent() const;
+    void setEvent(af_event event);
+    af_buffer_info get() const;
+    af_event unlockEvent();
+    void* unlockPtr();
+
+   private:
+    buffer_info& operator=(const buffer_info& other);
+    buffer_info(const buffer_info& other);
+};
+
+}  // namespace af
+#endif  // __cplusplus
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+AFAPI af_err af_create_buffer_info(af_buffer_info* buf, void* ptr,
+                                   af_event event);
+
+/// \brief deletes the \ref af_buffer_info and the resources its tracking
+///
+///  Deletes the \ref af_buffer_info object and its tracked resources. If buffer
+///  still holds
+/// the pointer,  that pointer is freed after its associated event has
+/// triggered.
+///
+/// \param[in] buf The af_buffer_info object that will be deleted
+/// \returns AF_SUCCESS
+AFAPI af_err af_delete_buffer_info(af_buffer_info buf);
+
+AFAPI af_err af_buffer_info_get_ptr(void** ptr, af_buffer_info buf);
+
+AFAPI af_err af_buffer_info_get_event(af_event* event, af_buffer_info buf);
+
+AFAPI af_err af_buffer_info_set_ptr(af_buffer_info buf, void* ptr);
+
+AFAPI af_err af_buffer_info_set_event(af_buffer_info buf, af_event event);
+
+/// \brief Disassociates the \ref af_event from the \ref af_buffer_info object
+///
+/// Gets the \ref af_event and disassociated it from the af_buffer_info object.
+/// Deleting the af_buffer_info object will not affect this event.
+///
+/// param[out] event The \ref af_event that will be disassociated. If NULL no
+/// event is
+///                   returned and the event is NOT freed
+/// param[in] buf The target \ref af_buffer_info object
+/// \returns AF_SUCCESS
+AFAPI af_err af_unlock_buffer_info_event(af_event* event, af_buffer_info buf);
+
+/// \brief Disassociates the pointer from the \ref af_buffer_info object
+///
+/// Gets the pointer and disassociated it from the \ref af_buffer_info object.
+/// Deleting the af_buffer_info object will not affect this pointer.
+///
+/// param[out] event The \ref pointer that will be disassociated. If NULL no
+/// pointer is
+///                  returned and the data is NOT freed.
+/// param[in] buf The target \ref af_buffer_info object
+/// \returns AF_SUCCESS
+AFAPI af_err af_unlock_buffer_info_ptr(void** ptr, af_buffer_info buf);
+
+////////////////////////////////////////////////////////////////////////////////
+// Memory Manager API
+////////////////////////////////////////////////////////////////////////////////
+
+typedef void (*af_memory_manager_initialize_fn)(af_memory_manager);
+
+typedef void (*af_memory_manager_shutdown_fn)(af_memory_manager);
+
+typedef af_buffer_info (*af_memory_manager_alloc_fn)(af_memory_manager, size_t,
```

Comment:
To summarize on some of my offline conversation with @umar456 — I'm making a few last changes to the API (wrapping things up!) to make it more general, including:
- Removing memory step size functions from the public memory manager API. Memory step size is an artifact of the default memory manager and not something all memory managers will implement. Since `af::[get/set]MemStepSize` it's part of the public API, we need to still support its functionality with the default memory manager. The new behavior is as follows:
  - By default, the public memory manager API behavior is unchanged.
  - Calls to public API step size functions throw an exception when a custom memory manager is set.
- All memory manager API functions now return an `af_err` and have pointers to return values passed in as arguments. This is to ensure ABI compatibility with how errors are surfaced and to ensure that we communicate across C/C++ runtimes with values rather than exceptions.
- `garbageCollect` is renamed throughout to `signalMemoryCleanup` to unopinionate naming

The above are already added to this PR. Things still to do include:
- Pass array shape, size, and type (via the number of bytes per element) to `alloc` by adding arguments `const unsigned ndims, const dim_t * const dims, const unsigned elementSize`
- Finalize the new `memoryPressure` metric so that we can remove JIT calls to `get_max_bytes`, `get_max_buffers`, `usage_info`, `check_memory_limit`, etc. These will be replaced by the following API:
  - `getMemoryPressure()` which returns a float between 0 and 1
  - `getMemoryPressureThreshold()` which lives on the memory manager and returns the threshold over which the JIT eval should be performed (there is sufficient memory pressure)
  - `setMemoryPressureThreshold()`
- Documentation

Reply:
[added]